README.md
<p align="center">
    <img alt="react-native-country-picker-modal" src="https://thumbs.gfycat.com/HandsomeInnocentAnura-size_restricted.gif" width=150>
</p>

<h3 align="center">
  The best Country Picker for React Native.
</h3>

<p align="center">
  <a href="https://reactnative.gallery"><img src="https://img.shields.io/badge/reactnative.gallery-%E2%99%A5-red.svg"></a>
  <a href="https://www.npmjs.com/package/react-native-country-picker-modal"><img src="https://img.shields.io/npm/v/react-native-country-picker-modal.svg?style=flat-square"></a>
  <a href="https://www.npmjs.com/package/react-native-country-picker-modal"><img src="https://img.shields.io/npm/dm/react-native-country-picker-modal.svg?style=flat-square"></a>
  <a href="https://codecov.io/gh/xcarpentier/react-native-country-picker-modal"><img src="https://codecov.io/gh/xcarpentier/react-native-country-picker-modal/coverage.svg"></a>
  <a href="https://circleci.com/gh/xcarpentier/react-native-country-picker-modal"><img src="https://circleci.com/gh/xcarpentier/react-native-country-picker-modal.svg?style=svg"></a>
</p>

<p align="center">
  <img src="https://api.qrserver.com/v1/create-qr-code/?size=100x100&data=https://exp.host/@xcarpentier/react-native-country-picker-modal">
  <br>
  <a href="https://exp.host/@xcarpentier/react-native-country-picker-modal">Demo on Expo</a>
</p>

<br />

## Installation
<<<<<<< HEAD

=======
>>>>>>> c4f1ed8e1e8519f701b08b2a40944fbbeef0a503
```bash
$ yarn add react-native-country-picker-modal
```

## Basic Usage
<<<<<<< HEAD

=======
>>>>>>> c4f1ed8e1e8519f701b08b2a40944fbbeef0a503
```jsx
import DeviceInfo from 'react-native-device-info'

import React, {
  AppRegistry,
  Component,
  StyleSheet,
  Text,
  View,
  StatusBarIOS,
  PixelRatio
} from 'react-native'
import CountryPicker, {
  getAllCountries
} from 'react-native-country-picker-modal'

const NORTH_AMERICA = ['CA', 'MX', 'US']

class Example extends Component {
  constructor(props) {
    StatusBarIOS.setHidden(true)
    super(props)
    let userLocaleCountryCode = DeviceInfo.getDeviceCountry()
    const userCountryData = getAllCountries()
      .filter(country => NORTH_AMERICA.includes(country.cca2))
      .filter(country => country.cca2 === userLocaleCountryCode)
      .pop()
    let callingCode = null
    let cca2 = userLocaleCountryCode
    if (!cca2 || !userCountryData) {
      cca2 = 'US'
      callingCode = '1'
    } else {
      callingCode = userCountryData.callingCode
    }
    this.state = {
      cca2,
      callingCode
    }
  }
  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.welcome}>Welcome to Country Picker !</Text>
        <CountryPicker
          countryList={NORTH_AMERICA}
          onChange={value => {
            this.setState({ cca2: value.cca2, callingCode: value.callingCode })
          }}
          cca2={this.state.cca2}
          translation="eng"
        />
        <Text style={styles.instructions}>press on the flag</Text>
        {this.state.country && (
          <Text style={styles.data}>
            {JSON.stringify(this.state.country, null, 2)}
          </Text>
        )}
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10
  },
  instructions: {
    fontSize: 12,
    textAlign: 'center',
    color: '#888',
    marginBottom: 5
  },
  data: {
    padding: 15,
    marginTop: 10,
    backgroundColor: '#ddd',
    borderColor: '#888',
    borderWidth: 1 / PixelRatio.get(),
    color: '#777'
  }
})

AppRegistry.registerComponent('example', () => Example)
```

## Props

| Key               | Type     | Default                                                                                                      | Description                                                                                                                           |
| ----------------- | -------- | ------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------- |
| cca2              | string   | \*required                                                                                                   | code ISO 3166-1 alpha-2 (ie. FR, US, etc.)                                                                                            |
| translation       | string   | 'eng'                                                                                                        | The language display for the name of the country (deu, fra, hrv, ita, jpn, nld, por, rus, spa, svk, fin, zho, cym)                    |
| onChange          | function | \*required                                                                                                   | The handler when a country is selected                                                                                                |
| onClose           | function | \*required                                                                                                   | The handler when the close button is clicked                                                                                          |
| countryList       | array    | See [cca2.json](https://github.com/xcarpentier/react-native-country-picker-modal/blob/master/data/cca2.json) | List of custom CCA2 countries to render in the list. Use getAllCountries to filter what you need if you want to pass in a custom list |
| excludeCountries  | array    | []                                                                                                           | List of custom CCA2 countries you don't want to render                                                                                |
| closeable         | bool     | false                                                                                                        | If true, the CountryPicker will have a close button                                                                                   |
| filterable        | bool     | false                                                                                                        | If true, the CountryPicker will have search bar                                                                                       |
| filterPlaceholder | string   | 'Filter'                                                                                                     | The search bar placeholder                                                                                                            |
| autoFocusFilter   | bool     | true                                                                                                         | Whether or not the search bar should be autofocused                                                                                   |
| styles            | object   | {}                                                                                                           | Override any style specified in the component (see source code)                                                                       |
| disabled          | bool     | false                                                                                                        | Whether or not the Country Picker onPress is disabled                                                                                 |
| tranparent        | bool     | false                                                                                                        | If true, the CountryPicker will render the modal over a transparent background                                                        |                                                                                                              
| animationType     | string   |'none'                                                                                                        | The handler that controls how the modal animates                                                                                      |
| closeButtonImage  | React.element| default close button Image                                                                               | Custom close button Image

## Dependencies

* world-countries : https://www.npmjs.com/package/world-countries

## FAQ

### Is it supported and tested both on android and iOS?

YES

### Is the data that is populated inside the list saved offline once I install your package?

YES : It used the world-countries package and image is stored into json and base64.

## Tiers lib using this lib

* [react-native-phone-verification](https://github.com/joinspontaneous/react-native-phone-verification)

[> Your project?](https://github.com/xcarpentier/react-native-linkedin/issues/new)

## See also

* [react-native-linkedin](https://github.com/xcarpentier/react-native-linkedin)

## Contribution

* [@xcapentier](mailto:contact@xaviercarpentier.com) The main author.

## Questions

Feel free to [contact me](mailto:contact@xaviercarpentier.com) or [create an issue](https://github.com/xcarpentier/react-native-country-picker/issues/new)

> made with ♥

## Licence

[MIT](https://github.com/xcarpentier/react-native-country-picker-modal/blob/master/LICENSE.md)


CHANGELOG.md
<a name="0.5.1"></a>
## [0.5.1](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.5.0...v0.5.1) (2017-12-11)


### Features

* **openModal:**  added in constructor of CountryPicker.js ([e1d234c](https://github.com/xcarpentier/react-native-country-picker-modal/commit/e1d234c))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.4.0...v0.5.0) (2017-12-06)



<a name="0.4.0"></a>
# [0.4.0](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.3.0...v0.4.0) (2017-11-20)


### Features

* **expo:** add expo example ([5ab2eba](https://github.com/xcarpentier/react-native-country-picker-modal/commit/5ab2eba))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.2.7...v0.3.0) (2017-06-30)


### Features

* **filter:** merge ([5890470](https://github.com/xcarpentier/react-native-country-picker-modal/commit/5890470))
* **jest:** add jest ([d2b6857](https://github.com/xcarpentier/react-native-country-picker-modal/commit/d2b6857))
* **search:** add search bar ([#55](https://github.com/xcarpentier/react-native-country-picker-modal/issues/55)) ([c29f27a](https://github.com/xcarpentier/react-native-country-picker-modal/commit/c29f27a))



<a name="0.2.7"></a>
## [0.2.7](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.2.6...v0.2.7) (2016-09-28)


### Features

* **flags:** emoji ([2fa49bd](https://github.com/xcarpentier/react-native-country-picker-modal/commit/2fa49bd))
* **ios:** use emoji ([655df40](https://github.com/xcarpentier/react-native-country-picker-modal/commit/655df40))
* **ios:** use emoji ([64580b3](https://github.com/xcarpentier/react-native-country-picker-modal/commit/64580b3))



<a name="0.2.6"></a>
## [0.2.6](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.2.5...v0.2.6) (2016-09-26)


### Bug Fixes

* **babel:** preset on dependencies ([bf952a9](https://github.com/xcarpentier/react-native-country-picker-modal/commit/bf952a9))



<a name="0.2.5"></a>
## [0.2.5](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.2.3...v0.2.5) (2016-09-21)


### Features

* **data:** reduce data size ([#29](https://github.com/xcarpentier/react-native-country-picker-modal/issues/29)) ([ed9a301](https://github.com/xcarpentier/react-native-country-picker-modal/commit/ed9a301))



<a name="0.2.3"></a>
## [0.2.3](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.2.2...v0.2.3) (2016-09-07)


### Bug Fixes

* **lang:** add zho ([0541520](https://github.com/xcarpentier/react-native-country-picker-modal/commit/0541520))


### Features

* **open modal:** from outside via ref ([f0c8e0a](https://github.com/xcarpentier/react-native-country-picker-modal/commit/f0c8e0a))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/xcarpentier/react-native-country-picker-modal/compare/v0.1.6...v0.2.1) (2016-09-05)



<a name="0.1.6"></a>
## [0.1.6](https://github.com/xcarpentier/react-native-country-picker-modal/compare/0.0.1...v0.1.6) (2016-07-20)


### Bug Fixes

* **Dimension:** the dimensions error still seems to be an issue ([ff8dd76](https://github.com/xcarpentier/react-native-country-picker-modal/commit/ff8dd76))


### Features

* **closeable:** inccrease version ([28cd44c](https://github.com/xcarpentier/react-native-country-picker-modal/commit/28cd44c))
* **example:** adding example ([a8aeccb](https://github.com/xcarpentier/react-native-country-picker-modal/commit/a8aeccb))
* **example:** better font style ([a001bab](https://github.com/xcarpentier/react-native-country-picker-modal/commit/a001bab))





src/CountryPicker.js
// @flow
/* eslint import/newline-after-import: 0 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import {
  StyleSheet,
  View,
  Image,
  TouchableOpacity,
  Modal,
  Text,
  TextInput,
  ListView,
  ScrollView,
  Platform
} from 'react-native'

import Fuse from 'fuse.js'

import cca2List from '../data/cca2'
import { getHeightPercent } from './ratio'
import CloseButton from './CloseButton'
import countryPickerStyles from './CountryPicker.style'
import KeyboardAvoidingView from './KeyboardAvoidingView'

let countries = null
let Emoji = null
let styles = {}

let isEmojiable = Platform.OS === 'ios'

const FLAG_TYPES = {
  flat: 'flat',
  emoji: 'emoji'
}

const setCountries = flagType => {
  if (typeof flagType !== 'undefined') {
    isEmojiable = flagType === FLAG_TYPES.emoji
  }

  if (isEmojiable) {
    countries = require('../data/countries-emoji')
    Emoji = require('./emoji').default
  } else {
    countries = require('../data/countries')
    Emoji = <View />
  }
}

const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 })

setCountries()

export const getAllCountries = () =>
  cca2List.map(cca2 => ({ ...countries[cca2], cca2 }))

export default class CountryPicker extends Component {
  static propTypes = {
    cca2: PropTypes.string.isRequired,
    translation: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    onClose: PropTypes.func,
    closeable: PropTypes.bool,
    filterable: PropTypes.bool,
    children: PropTypes.node,
    countryList: PropTypes.array,
    excludeCountries: PropTypes.array,
    styles: PropTypes.object,
    filterPlaceholder: PropTypes.string,
    autoFocusFilter: PropTypes.bool,
    // to provide a functionality to disable/enable the onPress of Country Picker.
    disabled: PropTypes.bool,
    filterPlaceholderTextColor: PropTypes.string,
    closeButtonImage: PropTypes.element,
    transparent: PropTypes.bool,
    animationType: PropTypes.oneOf(['slide', 'fade', 'none']),
    flagType: PropTypes.oneOf(Object.values(FLAG_TYPES)),
    hideAlphabetFilter: PropTypes.bool
  }

  static defaultProps = {
    translation: 'eng',
    countryList: cca2List,
    excludeCountries: [],
    filterPlaceholder: 'Filter',
    autoFocusFilter: true,
    transparent: false,
    animationType: 'none'
  }

  static renderEmojiFlag(cca2, emojiStyle) {
    return (
      <Text style={[styles.emojiFlag, emojiStyle]} allowFontScaling={false}>
        {cca2 !== '' && countries[cca2.toUpperCase()] ? (
          <Emoji name={countries[cca2.toUpperCase()].flag} />
        ) : null}
      </Text>
    )
  }

  static renderImageFlag(cca2, imageStyle) {
    return cca2 !== '' ? (
      <Image
        style={[styles.imgStyle, imageStyle]}
        source={{ uri: countries[cca2].flag }}
      />
    ) : null
  }

  static renderFlag(cca2, itemStyle, emojiStyle, imageStyle) {
    return (
      <View style={[styles.itemCountryFlag, itemStyle]}>
        {isEmojiable
          ? CountryPicker.renderEmojiFlag(cca2, emojiStyle)
          : CountryPicker.renderImageFlag(cca2, imageStyle)}
      </View>
    )
  }

  constructor(props) {
    super(props)
    this.openModal = this.openModal.bind(this)

    setCountries(props.flagType)
    let countryList = [...props.countryList]
    const excludeCountries = [...props.excludeCountries]

    excludeCountries.forEach(excludeCountry => {
      const index = countryList.indexOf(excludeCountry)

      if (index !== -1) {
        countryList.splice(index, 1)
      }
    })

    // Sort country list
    countryList = countryList
      .map(c => [c, this.getCountryName(countries[c])])
      .sort((a, b) => {
        if (a[1] < b[1]) return -1
        if (a[1] > b[1]) return 1
        return 0
      })
      .map(c => c[0])

    this.state = {
      modalVisible: false,
      cca2List: countryList,
      dataSource: ds.cloneWithRows(countryList),
      filter: '',
      letters: this.getLetters(countryList)
    }

    if (this.props.styles) {
      Object.keys(countryPickerStyles).forEach(key => {
        styles[key] = StyleSheet.flatten([
          countryPickerStyles[key],
          this.props.styles[key]
        ])
      })
      styles = StyleSheet.create(styles)
    } else {
      styles = countryPickerStyles
    }

    this.fuse = new Fuse(
      countryList.reduce(
        (acc, item) => [
          ...acc,
          { id: item, name: this.getCountryName(countries[item]) }
        ],
        []
      ),
      {
        shouldSort: true,
        threshold: 0.6,
        location: 0,
        distance: 100,
        maxPatternLength: 32,
        minMatchCharLength: 1,
        keys: ['name'],
        id: 'id'
      }
    )
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.countryList !== this.props.countryList) {
      this.setState({
        cca2List: nextProps.countryList,
        dataSource: ds.cloneWithRows(nextProps.countryList)
      })
    }
  }

  onSelectCountry(cca2) {
    this.setState({
      modalVisible: false,
      filter: '',
      dataSource: ds.cloneWithRows(this.state.cca2List)
    })

    this.props.onChange({
      cca2,
      ...countries[cca2],
      flag: undefined,
      name: this.getCountryName(countries[cca2])
    })
  }

  onClose() {
    this.setState({
      modalVisible: false,
      filter: '',
      dataSource: ds.cloneWithRows(this.state.cca2List)
    })
    if (this.props.onClose) {
      this.props.onClose()
    }
  }

  getCountryName(country, optionalTranslation) {
    const translation = optionalTranslation || this.props.translation || 'eng'
    return country.name[translation] || country.name.common
  }

  setVisibleListHeight(offset) {
    this.visibleListHeight = getHeightPercent(100) - offset
  }

  getLetters(list) {
    return Object.keys(
      list.reduce(
        (acc, val) => ({
          ...acc,
          [this.getCountryName(countries[val])
            .slice(0, 1)
            .toUpperCase()]: ''
        }),
        {}
      )
    ).sort()
  }

  openModal = this.openModal.bind(this)

  // dimensions of country list and window
  itemHeight = getHeightPercent(7)
  listHeight = countries.length * this.itemHeight

  openModal() {
    this.setState({ modalVisible: true })
  }

  scrollTo(letter) {
    // find position of first country that starts with letter
    const index = this.state.cca2List
      .map(country => this.getCountryName(countries[country])[0])
      .indexOf(letter)
    if (index === -1) {
      return
    }
    let position = index * this.itemHeight

    // do not scroll past the end of the list
    if (position + this.visibleListHeight > this.listHeight) {
      position = this.listHeight - this.visibleListHeight
    }

    // scroll
    this._listView.scrollTo({
      y: position
    })
  }

  handleFilterChange = value => {
    const filteredCountries =
      value === '' ? this.state.cca2List : this.fuse.search(value)

    this._listView.scrollTo({ y: 0 })

    this.setState({
      filter: value,
      dataSource: ds.cloneWithRows(filteredCountries)
    })
  }

  renderCountry(country, index) {
    return (
      <TouchableOpacity
        key={index}
        onPress={() => this.onSelectCountry(country)}
        activeOpacity={0.99}
      >
        {this.renderCountryDetail(country)}
      </TouchableOpacity>
    )
  }

  renderLetters(letter, index) {
    return (
      <TouchableOpacity
        key={index}
        onPress={() => this.scrollTo(letter)}
        activeOpacity={0.6}
      >
        <View style={styles.letter}>
          <Text style={styles.letterText} allowFontScaling={false}>
            {letter}
          </Text>
        </View>
      </TouchableOpacity>
    )
  }

  renderCountryDetail(cca2) {
    const country = countries[cca2]
    return (
      <View style={styles.itemCountry}>
        {CountryPicker.renderFlag(cca2)}
        <View style={styles.itemCountryName}>
          <Text style={styles.countryName} allowFontScaling={false}>
            {this.getCountryName(country)}
          </Text>
        </View>
      </View>
    )
  }

  render() {
    return (
      <View>
        <TouchableOpacity
          disabled={this.props.disabled}
          onPress={() => this.setState({ modalVisible: true })}
          activeOpacity={0.7}
        >
          {this.props.children ? (
            this.props.children
          ) : (
            <View
              style={[styles.touchFlag, { marginTop: isEmojiable ? 0 : 5 }]}
            >
              {CountryPicker.renderFlag(this.props.cca2)}
            </View>
          )}
        </TouchableOpacity>
        <Modal
          transparent={this.props.transparent}
          animationType={this.props.animationType}
          visible={this.state.modalVisible}
          onRequestClose={() => this.setState({ modalVisible: false })}
        >
          <View style={styles.modalContainer}>
            <View style={styles.header}>
              {this.props.closeable && (
                <CloseButton
                  image={this.props.closeButtonImage}
                  styles={[styles.closeButton, styles.closeButtonImage]}
                  onPress={() => this.onClose()}
                />
              )}
              {this.props.filterable && (
                <TextInput
                  autoFocus={this.props.autoFocusFilter}
                  autoCorrect={false}
                  placeholder={this.props.filterPlaceholder}
                  placeholderTextColor={this.props.filterPlaceholderTextColor}
                  style={[
                    styles.input,
                    !this.props.closeable && styles.inputOnly
                  ]}
                  onChangeText={this.handleFilterChange}
                  value={this.state.filter}
                  allowFontScaling={false}
                />
              )}
            </View>
            <KeyboardAvoidingView behavior="padding">
              <View style={styles.contentContainer}>
                <ListView
                  keyboardShouldPersistTaps="always"
                  enableEmptySections
                  ref={listView => (this._listView = listView)}
                  dataSource={this.state.dataSource}
                  renderRow={country => this.renderCountry(country)}
                  initialListSize={30}
                  pageSize={15}
                  onLayout={({ nativeEvent: { layout: { y: offset } } }) =>
                    this.setVisibleListHeight(offset)
                  }
                />
                {!this.props.hideAlphabetFilter && (
                  <ScrollView
                    contentContainerStyle={styles.letters}
                    keyboardShouldPersistTaps="always"
                  >
                    {this.state.filter === '' &&
                      this.state.letters.map((letter, index) =>
                        this.renderLetters(letter, index)
                      )}
                  </ScrollView>
                )}
              </View>
            </KeyboardAvoidingView>
          </View>
        </Modal>
      </View>
    )
  }
}
